### ISSUE TITLE
Overlay items flicker when near the edge of the window

### ISSUE THREAD

**Issue by ssfrr** [atom/atom#8295]
This happens frequently when using the [Hydrogen package](https://github.com/willwhitney/hydrogen), which displays the return values from evaluated code as overlays on the window. It seems to be exacerbated when there are more overlays. It looks like the layout engine is trying to figure out where to render them but the act of rendering changes the location solution, so they start flickering like crazy.

Originally filed as an issue on Hydrogen: willwhitney/hydrogen#93, but it seems this is an issue with the Atom rendering.

**Comment by izuzak** [comment]
@ssfrr Just to make sure we have all the relevant information -- can you please record a short GIF which clearly demonstrates the problem, and also provide the exact steps which reproduce this for you (+ version of Atom and OS)?

**Comment by ssfrr** [comment]
This gif doesn't really capture the speed of the flicker (it's very fast). Video is [here] (https://vimeo.com/136067838) (and even that seems slower then it looks on my screen).

![atom_flicker](https://cloud.githubusercontent.com/assets/734386/9216613/1497f656-408a-11e5-8f3e-fb3ea649ad81.gif)

I'm on OSX, using Atom 1.0.5.

It seems to take a while to reproduce, and seems to be exacerbated when I have a lot of overlay items, which happens commonly after working in Hydrogen for a while. Unfortunately I don't have a specific sequence of steps, but if I start to see some commonalities I'll post back.

**Comment by simurai** [comment]
I couldn't get the Hydrogen package to install, but based on the video, it looks like some GPU layer issue.

If you're familiar with the DevTools, you could try to "Inspect Elements" and see what class that overlay has. It might be [.hydrogen.output-bubble](https://github.com/willwhitney/hydrogen/blob/master/styles/hydrogen.less#L14). If so, here some random things you can add to your `styles.less` and see if it helps.

```less

// uncomment only one line at the time, not all together

.hydrogen.output-bubble,
atom-text-editor::shadow .hydrogen.output-bubble {
  //backface-visibility: hidden;
  //z-index: 0; // also try increasing the number by one: 1, 2, 3 etc.
  //will-change: transform;
}

atom-text-editor {
  //overflow: hidden; // this might cause autocomplete to be cut off
}
```

**Comment by izuzak** [comment]
Thanks, @ssfrr. 

@nathansobo Does the behavior shown in the GIF above ring any :bell: for you?

**Comment by ssfrr** [comment]
Here's what I see in the inspector. This was without any visible flickering on the screen, and no motion (scrolling or cursor movements) happening.
![inspector_flickering](https://cloud.githubusercontent.com/assets/734386/9232804/bc9ff286-40fd-11e5-942e-a7272813b8bb.gif)

Also my CPU was pegged at 100% while I had the inspector open.

**Comment by ssfrr** [comment]
@simurai the `hydrogen output-bubble` div is a child of the `atom-overlay` element, which is what seems to be flickering. I did try your suggestions though and did not see any change. After I made the changes I just saved my styles.less file and switched tabs back to the document with the flickering.

I also tried applying the styles to the atom-overlay element type as in:

```css
atom-overlay {
  // backface-visibility: hidden;
  z-index: 1; // also try increasing the number by one: 1, 2, 3 etc.
  // will-change: transform;
}
```

`will-change` had no effect. `z-index` had no effect for z-index > 0, if z-index == 0 the bubbles disappear entirely.
backface-visibility had no effect.

**Comment by simurai** [comment]
@ssfrr Thanks for testing.

Ok, based on the gif of the DevTools, it's __not__ a GPU rendering issue, but rather the `top` position of the `atom-overlay` keeps getting updated for no reason.

**Comment by ssfrr** [comment]
Sounds like a good clue :mag:. Let me know if there's anything else I can test to get you more info.

**Comment by nilshg** [comment]
Just to add that this is also a problem with the [atom-julia-client](https://github.com/JunoLab/atom-julia-client) package, where both results (Arrays resulting from computations) and autocomplete suggestions display this issue.

**Comment by as-cii** [comment]
Could you tell whether this was fixed in the recent versions? We have made a couple of changes in the way we position elements since this was first opened and I wonder if they have had any positive impact on this bug. If not, could you provide some reproduction steps so that I can jump on this?

Thanks!

**Comment by benogle** [comment]
I can reproduce this in ac+ when it's near the left of the window. This might be an issue with ac+ as it uses margin to adjust the left to line up the beginning of the words:

![ac flicker](https://cloud.githubusercontent.com/assets/69169/11821383/09d4f04c-a31e-11e5-9c4f-971bbf9d6921.gif)
_It's hard to capture with the gif as it's skipping frames..._

* close treeview and anything in the left panel
* type `atom.` and type some atom global that would complete like `clip`

**Comment by as-cii** [comment]
So, I can reproduce the above issue and have successfully debugged it. :mag_right:  It's caused by autocomplete-plus changing `SuggestionListElement` after reading UI properties, as you suggested. So, basically here's what's happening under the hood:

1. The user presses a keystroke, causing `SuggestionListElement` to render new elements.
2. A "poll after next update" is scheduled. (`Poll 1`)
3. A document update is scheduled. (`Update 1`)
4. A document read is scheduled. (`Read 1`)
5. `Update 1` is served and the overlay contents changed.
6. `Read 1` is served<br>
a) Some DOM properties are read<br>
b) A new document update is scheduled (`Update 2`)
7. `Poll 1` is served and overlays are measured. This causes a new update to `TextEditorComponent` be scheduled (`Update 3`)
8. `Update 2` is served and overlays are changed to adjust the margin.
9. `Update 3` is served, but overlays won't have the latest changes caused by `Update 2` because they won't be measured until the next document poll or read, which will happen on the next frame. (Flicker!)

A PR is on the way to [atom/autocomplete-plus](https://github.com/atom/autocomplete-plus), but I am curious if the other packages mentioned on this issue are experiencing the same problem. 

@ssfrr @nilshg: is there a way I can consistently trigger that behavior on your packages? I am trying to find commonalities between the AC+ flicker and the ones you're experiencing so that we could (maybe) provide a more holistic solution.

Thanks! :sparkles:

**Comment by ssfrr** [comment]
This doesn't seem to be a problem any more for me (using the atom-julia-client package on Atom 1.3.2). 

Whatever rendering change were made seem to help things stay a lot more stable. Thanks for the investigation!

**Comment by nilshg** [comment]
Unfortunately I'm still seeing this (Atom 1.3.3 on Win7 with master of atom-julia-client/atom-ink and all the .jl packages).

**Comment by benogle** [comment]
Can someone who is still experiencing this please post a gif and specific steps to repro?

**Comment by ypresto** [comment]
1. Install [merge-conflicts](https://github.com/smashwilson/merge-conflicts) on Atom 1.6.2
2. Execute merge-conflicts:detect command
3. Adjust window width
4. It starts extending to outside of left edge

![147dff1cfb85d28c572fe37d4464c724](https://cloud.githubusercontent.com/assets/400558/14320828/81dfd4c6-fc51-11e5-9f75-43ef4b38f072.gif)

**Comment by ypresto** [comment]
This plugin does:

```
  attached: ->
    view = atom.views.getView(@editor)
    @parent().css right: view.getVerticalScrollbarWidth()
```

where `@parent` is overlay.
when right is not specified it does not reproduce.

https://github.com/atom/atom/blob/10acfd057fbf994fe2ddd5dd12d66b7dba5cee94/src/text-editor-presenter.coffee#L468

This code seems to calculate overflow width and move view to left.
But `@windowWidth` is different than actual width so calculation does not converge.

**Comment by ypresto** [comment]
![image](https://cloud.githubusercontent.com/assets/400558/14326858/c26494c8-fc6a-11e5-8273-69114e815595.png)

omg...

**Comment by ypresto** [comment]
![image](https://cloud.githubusercontent.com/assets/400558/14326879/d66e6c64-fc6a-11e5-83e3-d79abb9d2fae.png)

omg...

**Comment by jeremyfa** [comment]
Having the same issue on atom 1.8.0, on OS X El Capitan. The issue occurs with autocomplete-plus, but also any cursor marker decoration of overlay type: https://github.com/atom/autocomplete-plus/issues/738

Any update on this issue?

**Comment by ypresto** [comment]
PR is here..! #11388

**Comment by nathansobo** [comment]
This will be fixed in 1.9.6 thanks to @ypresto! ❤️  ⚡

**Comment by jeremyfa** [comment]
The flickering issue is still happening on atom 1.9.6. I am not sure if what @ypresto fixed is the problem mentioned here initially. It seems that the problem he had (seeing his capture) is not the flickering issue described by @ssfrr .

On my side, I have the exact same problem as @ssfrr as he described here one year ago.

Please tell me if there is any other information I can provide to help get this fixed, as this is a rather annoying bug (like anything flickering on the screen when calmly writing code :( )

**Comment by ypresto** [comment]
I experienced similar behavior to this (position of overlay continuously being changed in inspector) https://github.com/atom/atom/issues/8295#issuecomment-130401143
So I thought this is same issue of mine, but it can be separated issue if it is not solved.

**Comment by ypresto** [comment]
Still experiencing related issue:

1. Install [merge-conflicts](https://github.com/smashwilson/merge-conflicts) on Atom 1.10.1
2. Execute merge-conflicts:detect command
3. Open or change tab to Project Find Results
4. Back to conflict-detected file
5. Overlay slowly shrinks from outside of left edge, then stop

(Show tree-view before step 3 for easy reproduction)

**Comment by ypresto** [comment]
While I could not reproduce, flickering issue similar to one @ssfrr reported happened.

![atom-bug-flicker](https://cloud.githubusercontent.com/assets/400558/18460775/5f2a8184-79af-11e6-845d-72e5f7298a15.gif)

(Low fps; it was much more faster than this gif.)

**Comment by nathansobo** [comment]
@ypresto @jeremyfa Could you let us know what themes you are using? We're having trouble reproducing.

**Comment by Ben3eeE** [comment]
I can get the autocomplete box to flicker when resizing panes but not without doing anything.

![flicker autocomplete 2](https://cloud.githubusercontent.com/assets/1058982/19326503/12a0f806-90cb-11e6-9362-77f81bd869ac.gif)

**Comment by nathansobo** [comment]
Closed via #12977.
